code/ai/aicode.cpp

```diff
@@ -6070,14 +6167,17 @@ int ai_fire_primary_weapon(object *objp)
 		enemy_sip = NULL;
 	}
 
-	if ( (swp->current_primary_bank < 0) || (swp->current_primary_bank >= swp->num_primary_banks) || timestamp_elapsed(aip->primary_select_timestamp)) {
+	//plieblang - added check for size of Preferred_primaries to force reevaluation if good-primary-time has been used in the meantime
+	if ( (swp->current_primary_bank < 0) || (swp->current_primary_bank >= swp->num_primary_banks) || timestamp_elapsed(aip->primary_select_timestamp) || Preferred_primary_info.size() != Preferred_primary_info_last_size) {
 		Weapon::Info_Flags flags = Weapon::Info_Flags::NUM_VALUES;
 		if ( aip->targeted_subsys != NULL ) {
 			flags = Weapon::Info_Flags::Puncture;
 		}
 		ai_select_primary_weapon(objp, enemy_objp, flags);
 		ship_primary_changed(shipp);	// AL: maybe send multiplayer information when AI ship changes primaries
 		aip->primary_select_timestamp = timestamp(5 * 1000);	//	Maybe change primary weapon five seconds from now.
+
+		Preferred_primary_info_last_size = Preferred_primary_info.size();
```

Comment:
If two *new* primaries are set, I think the code would be fine, but if existing preferences are changed, maybe not.  I think the best way to do this is to set the timestamp directly, since that's what the AI code usually expects.  I've made some modifications.